### PR TITLE
Unlock phpstan level 5

### PR DESCRIPTION
**Pull request by m-vo**
| Q                | A
| -----------------| ---
| Fixed issues     | -
| Docs PR or issue | -

I was annoyed by an invalid type annotation. But while fixing it I thought we could as well raise our phpstan level to 5. :smile: 

The commit messages should explain the motivation behind each set of changes.

### CODE REVIEW

File: core-bundle/src/Composer/ScriptHandler.php [hunk 1]
```diff
@@ -79,7 +79,7 @@ static function (string $type, string $buffer) use ($event): void {
         );
 
         if (!$process->isSuccessful()) {
-            throw new \RuntimeException(sprintf('An error occurred while executing the "%s" command.', $cmd));
+            throw new \RuntimeException(sprintf('An error occurred while executing the "%s" command.', implode(' ', $cmd)));
```
Comment (m-vo) [inline core-bundle/src/Composer/ScriptHandler.php:82]: (This was introduced with #1917 and therefore must not be backported.)

File: core-bundle/src/DataCollector/ContaoDataCollector.php [hunk 2]
```diff
@@ -145,7 +145,7 @@ private function addSummaryData(): void
             'framework' => $framework,
             'models' => $modelCount,
             'frontend' => isset($GLOBALS['objPage']),
-            'preview' => \defined('BE_USER_LOGGED_IN') && true === BE_USER_LOGGED_IN,
+            'preview' => \defined('BE_USER_LOGGED_IN') && true === \constant('BE_USER_LOGGED_IN'),
```
Comment (leofeyer) [inline core-bundle/src/DataCollector/ContaoDataCollector.php]: Why do we need `\constant()` here? We have already checked `\defined()`.

File: core-bundle/src/EventListener/MakeResponsePrivateListener.php [hunk 3]
```diff
@@ -53,7 +53,7 @@ public function __invoke(ResponseEvent $event): void
         $response = $event->getResponse();
 
         // Disable the default Symfony auto cache control
-        $response->headers->set(AbstractSessionListener::NO_AUTO_CACHE_CONTROL_HEADER, true);
+        $response->headers->set(AbstractSessionListener::NO_AUTO_CACHE_CONTROL_HEADER, '1');
```
Comment (leofeyer) [inline core-bundle/src/EventListener/MakeResponsePrivateListener.php:56]: Why `'1'`? FOSHttpCacheBundle uses `1` here: https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/master/src/EventListener/UserContextListener.php#L169

File: core-bundle/src/Image/ImageFactory.php [hunk 4]
```diff
@@ -112,6 +112,7 @@ public function setPredefinedSizes(array $predefinedSizes): void
 
     public function create($path, $size = null, $options = null): ImageInterface
     {
+        /** @var mixed $options */
```
Comment (leofeyer) [inline core-bundle/src/Image/ImageFactory.php]: Really? I figured `mixed` was the implied content type?

File: core-bundle/src/Resources/contao/library/Contao/Config.php [hunk 5]
```diff
@@ -373,8 +373,8 @@ public static function get($strKey)
 	/**
 	 * Temporarily set a configuration value
 	 *
-	 * @param string $strKey   The short key
-	 * @param string $varValue The configuration value
+	 * @param string     $strKey   The short key
+	 * @param mixed|null $varValue The configuration value
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Config.php]: Isn't `null` included in `mixed`?

File: core-bundle/src/Resources/contao/library/Contao/Picture.php [hunk 6]
```diff
@@ -64,7 +64,7 @@ class Picture
 	/**
 	 * The image size items collection
 	 *
-	 * @var ImageSizeItemModel[]|Collection
+	 * @var array<ImageSizeItemModel|object>|Collection
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Picture.php:67]: Where does `object` come from? Unit tests only or productional code?

File: core-bundle/src/Routing/UrlGenerator.php [hunk 7]
```diff
@@ -173,7 +177,7 @@ private function addHostToContext(RequestContext $context, array $parameters, in
         $context->setHost($host);
         $referenceType = UrlGeneratorInterface::ABSOLUTE_URL;
 
-        if (!$port) {
+        if (null === $port) {
```
Comment (leofeyer) [inline core-bundle/src/Routing/UrlGenerator.php]: Why this?

File: core-bundle/src/Routing/UrlGenerator.php [hunk 8]
```diff
@@ -186,13 +190,13 @@ private function addHostToContext(RequestContext $context, array $parameters, in
 
     /**
      * Extracts host and port from the domain.
-     *
-     * @return array<(string|null)>
```
Comment (leofeyer) [inline core-bundle/src/Routing/UrlGenerator.php:190]: It seems wrong to remove this entirely. Why not adjust it instead?

File: core-bundle/tests/Contao/ImageTest.php [hunk 9]
```diff
@@ -1001,8 +1001,6 @@ public function getGetLegacy(): \Generator
     public function testDoesNotFactorImagesInTheLegacyMethodIfTheArgumentIsInvalid(): void
     {
         $this->assertNull(Image::get('', 100, 100));
-        $this->assertNull(Image::get(0, 100, 100));
-        $this->assertNull(Image::get(null, 100, 100));
```
Comment (leofeyer) [inline core-bundle/tests/Contao/ImageTest.php:1005]: Why are these assertions no longer needed?

File: core-bundle/tests/EventListener/SubrequestCacheSubscriberTest.php [hunk 10]
```diff
@@ -42,14 +42,14 @@ public function testMergesCacheControlHeader(): void
         $this->onKernelRequest($subscriber, KernelInterface::MASTER_REQUEST);
 
         $subResponse = new Response();
-        $subResponse->headers->set(SubrequestCacheSubscriber::MERGE_CACHE_HEADER, true);
+        $subResponse->headers->set(SubrequestCacheSubscriber::MERGE_CACHE_HEADER, '1');
```
Comment (leofeyer) [inline core-bundle/tests/EventListener/SubrequestCacheSubscriberTest.php:45]: Again, why `'1'` instead of `1`?

File: core-bundle/tests/Image/ImageFactoryTest.php [hunk 11]
```diff
@@ -838,12 +838,6 @@ public static function emptyHookCallback(): void
     {
     }
 
-    /**
-     * @param ResizerInterface&MockObject $resizer
-     * @param ImagineInterface&MockObject $imagine
-     * @param ImagineInterface&MockObject $imagineSvg
-     * @param ContaoFramework&MockObject  $framework
-     */
```
Comment (leofeyer) [inline core-bundle/tests/Image/ImageFactoryTest.php]: Removing this will lead to `type MockObject is not within the annotated types` errors.

File: core-bundle/src/DataCollector/ContaoDataCollector.php (hunk 2 above)
Comment (m-vo) [reply core-bundle/src/DataCollector/ContaoDataCollector.php]: Because this way PHPStan won't complain about 'Strict comparison using === between true and false will always evaluate to false'. We ignored this previously but imo it doesn't hurt to use the constant function and rather have less exceptions.

File: core-bundle/src/EventListener/MakeResponsePrivateListener.php (hunk 3 above)
Comment (m-vo) [reply core-bundle/src/EventListener/MakeResponsePrivateListener.php:56]: :man_shrugging:  Because the `HeaderBag` expects values to be `string|string[]`.

see https://github.com/symfony/symfony/blob/73cb33a54e25e24b62844f971791e9765002422e/src/Symfony/Component/HttpFoundation/HeaderBag.php#L137

File: core-bundle/src/Image/ImageFactory.php (hunk 4 above)
Comment (m-vo) [reply core-bundle/src/Image/ImageFactory.php]: It's `int|array|ResizeConfiguration|null` (via `ImageFactoryInterface`). 
see https://github.com/contao/contao/blob/d8ae1290c620082105a0951a4f249c87590a3230/core-bundle/src/Image/ImageFactoryInterface.php#L26

We either must trust the annotation or remove the check. (Or we configure PHPStan to mistrust the annotated types - IIRC there is an option like this).

File: core-bundle/src/Resources/contao/library/Contao/Config.php (hunk 5 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/library/Contao/Config.php]: Yes it is, but the legacy code uses this notations all around. :slightly_smiling_face:  Shall I strip it for both occurrences in this class?

File: core-bundle/src/Resources/contao/library/Contao/Picture.php (hunk 6 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/library/Contao/Picture.php:67]: The class can operate on objects and the unit tests deliberately test it. But I'm also fine to rewrite the unit tests to use the models instead if we're sure this is no supported real world use case. 

/cc @ausi

File: core-bundle/src/Routing/UrlGenerator.php (hunk 7 above)
Comment (m-vo) [reply core-bundle/src/Routing/UrlGenerator.php]: Because now `$port` is a nullable `int`. Before it was a `string` / `empty string`.

File: core-bundle/src/Routing/UrlGenerator.php (hunk 8 above)
Comment (m-vo) [reply core-bundle/src/Routing/UrlGenerator.php:190]: Because now it returns a tuple `[string, ?int]` and PHP can not annotate those, yet. :smile: 

Do you prefer `@return array<(string|int|null)>`?

File: core-bundle/tests/Contao/ImageTest.php (hunk 9 above)
Comment (m-vo) [reply core-bundle/tests/Contao/ImageTest.php:1005]: Because they test with invalidly typed arguments.

To be more clear: The first parameter of `Image#get()` is annotated with `string`. If we consider this to be the "source of truth" for the API, then we should imo not test with `null` or `int`. (My reasoning is: This would fail if we'd upgrade to typed parameters.) If we *do* want to support these cases, then we could also loosen the type constraint and keep the tests (we could even add one with `false` because basically all that's tested is `$value == ''` :slightly_smiling_face:).

File: core-bundle/tests/Image/ImageFactoryTest.php (hunk 11 above)
Comment (m-vo) [reply core-bundle/tests/Image/ImageFactoryTest.php]: Where do you get this error?

File: core-bundle/tests/EventListener/SubrequestCacheSubscriberTest.php (hunk 10 above)
Comment (m-vo) [reply core-bundle/tests/EventListener/SubrequestCacheSubscriberTest.php:45]: see https://github.com/contao/contao/pull/1924#discussion_r452965031

File: core-bundle/src/Resources/contao/library/Contao/Picture.php (hunk 6 above)
Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/Picture.php:67]: I think objects are a supported use case. For hardcoded image size items you don’t need to create a model object (which was not easily possible in previous Contao versions AFAIK)

File: core-bundle/src/DataCollector/ContaoDataCollector.php (hunk 2 above)
Comment (ausi) [reply core-bundle/src/DataCollector/ContaoDataCollector.php]: I think I like ignoring this error better than using `constant()`

File: core-bundle/src/Image/ImageFactory.php (hunk 4 above)
Comment (ausi) [reply core-bundle/src/Image/ImageFactory.php]: Mistrusting the annotated types of method parameters seems reasonable to me.

File: core-bundle/src/Routing/UrlGenerator.php (hunk 7 above)
Comment (ausi) [reply core-bundle/src/Routing/UrlGenerator.php]: Port `0` would behave differently now, is that on purpose?

File: core-bundle/tests/Contao/ImageTest.php (hunk 9 above)
Comment (ausi) [reply core-bundle/tests/Contao/ImageTest.php:1005]: These tests should make sure that we keep BC, I would like to keep them. @m-vo could this stay if we configure phpstan to mistrust the function parameter annotated types?

File: core-bundle/src/Resources/contao/library/Contao/Image.php [hunk 12]
```diff
@@ -227,7 +227,7 @@ public function getImportantPart()
 	/**
 	 * Set the target height
 	 *
-	 * @param integer $targetHeight The target height
+	 * @param integer|float $targetHeight The target height
```
Comment (ausi) [inline core-bundle/src/Resources/contao/library/Contao/Image.php]: Why is this necessary?

File: core-bundle/src/Routing/UrlGenerator.php (hunk 7 above)
Comment (m-vo) [reply core-bundle/src/Routing/UrlGenerator.php]: Good question. IIRC you can bind to port 0 e.g. when using sockets (which means: connect to any free port) but it's not a valid TCP/UDP port per se. I'm not sure this edge case matters. But I can revert this change oc if you prefer.

File: core-bundle/src/DataCollector/ContaoDataCollector.php (hunk 2 above)
Comment (m-vo) [reply core-bundle/src/DataCollector/ContaoDataCollector.php]: Reverted in 1eb47e970be6eaa5a225104a7a7ba6f073b2bf63

Comment (m-vo) [reply core-bundle/src/DataCollector/ContaoDataCollector.php]: Found a better way: see https://phpstan.org/config-reference#constants

Changed in 63da7dd9541674f4f699d8a686cceab0b8c721b6

File: core-bundle/src/Resources/contao/library/Contao/Image.php (hunk 12 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/library/Contao/Image.php]: Because tests are inputting floats (and the method is casting the inputs to int). :see_no_evil:

Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/Image.php]: I think this is a similar topic that would be solved by mistrusting the annotated types.

File: core-bundle/tests/Contao/ImageTest.php (hunk 9 above)
Comment (m-vo) [reply core-bundle/tests/Contao/ImageTest.php:1005]: Had a quick discussion with @ausi about this topic. Probably PHPStan should ignore these cases where the type checking is done inside the method for legacy reasons (no real types in function signature + no `strict_types=1`) but it treats the annotations as certain.

I tried `treatPhpDocTypesAsCertain: false` but unfortunately this does not affect function parameters, so I'm going to ignore the errors for these few occurrences in our `phpstan.neon.dist` config.

File: core-bundle/src/Resources/contao/library/Contao/Config.php (hunk 5 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/library/Contao/Config.php]: Nevermind, I stripped it for both occurrences in a07ffbfb42c1d42f3f325c8be33e5c99a7bd6e04.

File: core-bundle/src/Resources/contao/library/Contao/Image.php (hunk 12 above)
Comment (m-vo) [reply core-bundle/src/Resources/contao/library/Contao/Image.php]: (This was also handled in daf0e0407d50b06b150e4997865c4629c4cdc71d.)

File: core-bundle/tests/Contao/ImageTest.php (hunk 9 above)
Comment (m-vo) [reply core-bundle/tests/Contao/ImageTest.php:1005]: Handled in daf0e0407d50b06b150e4997865c4629c4cdc71d.

File: core-bundle/src/Image/ImageFactory.php (hunk 4 above)
Comment (m-vo) [reply core-bundle/src/Image/ImageFactory.php]: As mistrusting types via config unfortunately [doesn't work](https://github.com/contao/contao/pull/1924#discussion_r453186771) in this case - what shall we do here?
1) explicitly mark parameter as uncertain `/** @var mixed $options */`
1) ignore checking for this statement `/** @phpstan-ignore-next-line */`
1) add exception to `phpstan.neon.dist`

I'd actually prefer the first option.

File: core-bundle/src/Routing/UrlGenerator.php (hunk 8 above)
Comment (m-vo) [reply core-bundle/src/Routing/UrlGenerator.php:190]: We could use psalms notation of object-like arrays `@return array{string, int|null}`, though this isn't even enforced in our current psalm level.

Comment (leofeyer) [reply core-bundle/src/Routing/UrlGenerator.php:190]: > Do you prefer `@return array<(string|int|null)>`?

I do.

File: core-bundle/tests/Image/ImageFactoryTest.php (hunk 11 above)
Comment (leofeyer) [reply core-bundle/tests/Image/ImageFactoryTest.php]: In PhpStorm. But I have just checked and it seems that it is no longer the case. 👍

Comment (leofeyer) [reply core-bundle/tests/Image/ImageFactoryTest.php]: See #1936.

File: core-bundle/src/Routing/UrlGenerator.php (hunk 8 above)
Comment (m-vo) [reply core-bundle/src/Routing/UrlGenerator.php:190]: Adressed in e3b80e01eb5868796b6b193829ac39756b304845.

File: core-bundle/src/Routing/UrlGenerator.php (hunk 7 above)
Comment (leofeyer) [reply core-bundle/src/Routing/UrlGenerator.php:180]: I am for reverting the change.

File: core-bundle/src/Routing/RouteProvider.php [hunk 13]
```diff
@@ -491,11 +490,12 @@ private function findPages(array $candidates): array
             return [];
         }
 
+        /** @var array<PageModel> */
```
Comment (leofeyer) [inline core-bundle/src/Routing/RouteProvider.php:493]: Shouldn't we add `@mixin<T>` and `@return array<T>` (or some other generics magic) to the `Model` class so PhpStan knows what we are returning here?

**Review by leofeyer — DISMISSED** [review]

File: core-bundle/src/Routing/UrlGenerator.php (hunk 7 above)
Comment (leofeyer) [reply core-bundle/src/Routing/UrlGenerator.php]: I am for reverting the change.

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/Routing/RouteProvider.php (hunk 13 above)
Comment (m-vo) [reply core-bundle/src/Routing/RouteProvider.php:493]: That's a tough one. We would imo need to add generics to the `Collection`, the `Adapter` and the `Model` for this to work. And then we will end with a lot of `@phpstan-param`, `@phpstan-return`, `@phpstan-var`... declarations so that our IDEs are still able to guess the type. 

Everything would look a bit like this I guess: https://phpstan.org/r/15377a79-1be1-4f2d-999f-f1ae9dbe37f0

File: core-bundle/src/Routing/UrlGenerator.php (hunk 7 above)
Comment (m-vo) [reply core-bundle/src/Routing/UrlGenerator.php]: Reverted in 77e2a4dc9e31bdb14f4a736ace119e5223052bc7.

File: core-bundle/src/Routing/RouteProvider.php (hunk 13 above)
Comment (leofeyer) [reply core-bundle/src/Routing/RouteProvider.php:493]: Do you want to try?

Comment (m-vo) [reply core-bundle/src/Routing/RouteProvider.php:493]: Challenge accepted. Are you ok with all those tags?

Comment (leofeyer) [reply core-bundle/src/Routing/RouteProvider.php:493]: Yes, I am. We are already using them in the [`Adapter`](https://github.com/contao/contao/blob/master/core-bundle/src/Framework/Adapter.php) class anyway.

Comment (m-vo) [reply core-bundle/src/Routing/RouteProvider.php:493]: I had a deep dive into the topic. 

Most of the things could *theoretically* be solved by having a `Collection<T of Model>` + `static`/`Collection<static>` return type annotations in the model base class and forced return overwrites for all of the `findBy*` methods because of the dynamic `Model` vs `Collection` vs `Array` return type nature of `find()`. 

Unfortunately, in reality this doesn't work as expected because we're unfortunately using `__callStatic` and PHPStan then is not able to resolve the types.

It would possibly work to create our own extension, but well...

Comment (bytehead) [reply core-bundle/src/Routing/RouteProvider.php:493]: Feel free to modify https://github.com/contao/phpstan 😎

Comment (m-vo) [reply core-bundle/src/Routing/RouteProvider.php:493]: > Feel free to modify https://github.com/contao/phpstan sunglasses 

If I only knew how. :smile:

**Comment by leofeyer** [comment]
Thank you @m-vo.

File: core-bundle/src/Routing/RouteProvider.php (hunk 13 above)
Comment (leofeyer) [reply core-bundle/src/Routing/RouteProvider.php:493]: I have created a follow-up issue here: #1940
